staging/src/k8s.io/apiserver/pkg/endpoints/handlers/rest.go

```diff
@@ -416,3 +420,28 @@ func parseTimeout(str string) time.Duration {
 func isDryRun(url *url.URL) bool {
 	return len(url.Query()["dryRun"]) != 0
 }
+
+type etcdError interface {
+	Code() grpccodes.Code
+	Error() string
+}
+
+type grpcError interface {
+	GRPCStatus() *grpcstatus.Status
+}
+
+func isTooLargeError(err error) bool {
+	if err != nil {
+		if etcdErr, ok := err.(etcdError); ok {
+			if etcdErr.Code() == grpccodes.InvalidArgument && etcdErr.Error() == "etcdserver: request is too large" {
+				return true
+			}
+		}
+		if grpcErr, ok := err.(grpcError); ok {
+			if grpcErr.GRPCStatus().Code() == grpccodes.ResourceExhausted && strings.Contains(grpcErr.GRPCStatus().Message(), "trying to send message larger than max") {
+				return true
+			}
+		}
```

Comment:
would like ack from @jpbetz on this

Reply:
@jpbetz PTAL, this is the last blocker for SSA (other than our perf issue).

Reply:
Yes, this is right.

```
head -c 1G </dev/urandom | bin/etcdctl put a
{"level":"warn","ts":"2020-02-14T16:25:50.671-0800","caller":"clientv3/retry_interceptor.go:61","msg":"retrying of unary invoker failed","target":"endpoint://client-b6802cd6-e6b5-4bd9-8c10-74c519f36bf9/127.0.0.1:2379","attempt":0,"error":"rpc error: code = ResourceExhausted desc = trying to send message larger than max (1073741833 vs. 2097152)"}
Error: rpc error: code = ResourceExhausted desc = trying to send message larger than max (1073741833 vs. 2097152)
```

The

Reply:
Actually, hold on, there is also: 

```
ErrGRPCRequestTooLarge        = status.New(codes.InvalidArgument, "etcdserver: request is too large").Err()
```

This is set by the etcd `--max-request-bytes` server flag.

```
head -c 1K </dev/urandom | bin/etcdctl put a
{"level":"warn","ts":"2020-02-14T16:34:30.737-0800","caller":"clientv3/retry_interceptor.go:61","msg":"retrying of unary invoker failed","target":"endpoint://client-6b986f0f-1f6e-4469-8a32-ed5370e4731c/127.0.0.1:2379","attempt":0,"error":"rpc error: code = InvalidArgument desc = etcdserver: request is too large"}
Error: etcdserver: request is too large
```

We need to check for both errors here.

xref: https://github.com/etcd-io/etcd/blob/425b65467c1facafc9605133cd1de4f1e7cfc3d8/etcdserver/api/v3rpc/rpctypes/error.go#L47, https://github.com/etcd-io/etcd/blob/c94782cd55fb44df43574505db9ac1c1b7d49c00/etcdserver/v3_server.go#L635

Reply:
@jpbetz that is checked in the line above: https://github.com/kubernetes/kubernetes/pull/87508/files#diff-390e9c6b4af51b3a3723c78f40b0e134R436

Reply:
excellent, I'd missed that